Add helper methods on primitive pointer types for pointer tagging (#3700)
rust-lang/rfcs: text/3700-ptr-tag-helpers.md

- Feature Name: `ptr_tag_helpers`
- Start Date: 2024-09-26
- RFC PR: [rust-lang/rfcs#3700](https://github.com/rust-lang/rfcs/pull/3700)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Add helper methods on primitive pointer types to facilitate getting and setting the tag of a pointer.

Reviewer: The term "tag of a pointer" means a lot of different things, and often it means something different from how you are using the term here. So the RFC should clarify the terminology it uses. This is, AFAIK, *not* meant to be a general pointer tagging mechanism. No RFC is even needed for that. It is specific for *working with hardware that ignores certain bits of a pointer*.

Author: The thinking is that hardware that ignores certain bits of a pointer is only relevant if you're setting those bits to some values, i.e. in the context of pointer tagging. Thus, the way to add support for such hardware is through a mechanism for pointer tagging. Does that reasoning make sense?

This is still a bit of an open question, there are several directions we could go with this. The tricky part is that different architectures have support for something similar, but the details (which exact bits are ignored) may vary.
On that account I'm not sure whether it'd be better to try and make a general high-bit pointer tagging mechanism that could be used across architectures, or whether it'd be better to just upstream e.g. aarch64-specific functions into core_arch and maybe then think about a generic one that just calls into those depending on the platform.
I don't really have very strong opinions here.

Reviewer: I'm just pointing out the RFC is written in a confusing way. I would suggest to make it very targeted specifically for "supporting hardware that ignores some bit in a pointer", rather than general pointer tagging. This should become clear already in the summary and the first paragraph of the motivation.

Intended to work with programs that make use of architecture features such as AArch64
Top-Byte Ignore (TBI), the primary use-case being writing tagging memory allocators.

# Motivation
[motivation]: #motivation

Tagged pointers are pointers in which the unused top bits are set to contain some metadata - the tag.
No 64-bit architecture today actually uses a 64-bit address space. Most operating systems only use
the lower 48 bits, leaving higher bits unused. The remaining bits are for the most part used to
distinguish userspace pointers (0x00) from kernelspace pointers (0xff).
Certain architectures provide extensions, such as TBI on AArch64, that allow programs to make use of

> those unused bits to insert custom metadata into the pointer.

Reviewer: To be clear, userspace can already do that without hardware support, trivially. By masking out those bots for each load. The hardware feature just makes this slightly more efficient.

I find the introduction to be a bit confusing due to this.

> Currently, Rust does not acknowledge TBI and related architecture extensions that enable the use of

Reviewer: The RFC should explain what the problem with the use of these extensions in Rust is.

Also, it is very odd to see interop with C/C++ as the motivation here, given that those languages *do not support TBI either* (which is another relevant fact the RFC should mention).

Author: Well they don't support it "fully" on a language level, as in you can't just tag arbitrary pointers with no issues, but they sort of do in that there are currently C/C++ components running in production that operate on TBI-enabled pointers.
I want to make it possible to, say, write a custom allocator which internally just calls malloc, puts some tag in the pointer it got from malloc and then returns that pointer to the user.

> Currently, Rust does not acknowledge TBI and related architecture extensions that enable the use of
> tagged pointers. This could potentially cause issues in cases such as working with TBI-enabled C/C++
> components over FFI, or when writing a tagging memory allocator.

Reviewer: ```suggestion
Currently, Rust does not support directly using TBI and related architecture extensions that simplify the use of
tagged pointers. This could potentially cause issues in cases such as working with TBI-enabled C/C++
components over FFI, or when writing a tagging memory allocator.
```

These functions are worth including in the standard library, despite their relatively niche use case
and relative simplicity, so that there is a single known location where Miri hooks can be called to
update the canonical address.
This will make it easier to modify tagged pointers without breaking the Rust memory model.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This RFC adds two methods on each primitive pointer type - `ptr.tag()` and
`ptr.with_tag(tag: u8)`.

```
assert!(ptr.tag() == 0);
let tagged_ptr = unsafe { ptr.with_tag(63) };

> assert!(tagged_ptr.tag() == 63);

Reviewer: The guide-level explanation should give a guide for how to think about these functions and what they do.

This is where you have to explain that `tag` is basically `realloc`. Currently, `realloc` suddenly appears in "drawbacks", which nobody reading the RFC will understand.

Remember that an RFC is supposed to be a self-contained document such that everyone who is reasonably well-versed in Rust but knows *nothing* about the particular problem domain can understand what problem the RFC is trying to solve, and how it is trying to solve it. Your RFC is missing a lot of background and explanation to make it satisfy this requirement.

```

The primary use-case is implementing an allocator that tags pointers before returning them to the
caller.

# Reference-level explanation

> [reference-level-explanation]: #reference-level-explanation

Reviewer: This section should contain the full signature of the newly proposed methods, and their doc comments (see the existing pointer methods for how our doc comments look like), so that we can an exact description of what it is they are doing.

Author: Yes for sure, the reason they're not there yet is that I wanted to ask for outside opinions before settling in on what those signatures should be and what the exact methods should actually look like. My bad that it didn't come across as intended.
The concrete part of the proposal is that we should have a function/method to set the high-bit tag of a pointer & one to retrieve it, the other details such as the name, where it should live or what it should entail are still in flux.

Within Rust's memory model, modifying the high bits offsets the pointer outside of the bounds of

> its original allocation, making any use of it Undefined Behaviour.

Reviewer: ```suggestion
its original allocation, making any load/store with that pointer Undefined Behaviour.
```
"any use" is not correct, e.g. `wrapping_offset` or `==` on such a pointer are completely fine.

The `with_tag()` method is only designed to be a helper for writing tagging allocators.
Users *must* ensure that code using this method simulates a realloc from the untagged
address to the tagged address, and that the underlying memory is only ever accessed
using the tagged address from there onwards. Anything short of that explicitly violates the
Rust memory model and will cause the program to break in unexpected ways.

# Drawbacks
[drawbacks]: #drawbacks

Because the memory model we currently have is not fully compatible with memory tagging and

> tagged pointers, setting the high bits of a pointer must be done with great care in order to

Reviewer: That's not correct. Pointer tagging works perfectly fine and is a commonly used technique in Rust. The memory model is just not compatible with *loading from a tagged pointer without first removing the tag*.

Author: That's why I wrote "not fully compatible" - you can make it work, just not as smoothly as one would like given what the hardware can support.
Besides, if you remove the tag, would that not make the result a different pointer? They won't compare equal after all.
Either way, I can clarify for sure if it's not self evident.
